### DISCUSSION TITLE
Style for base and derived class

### DISCUSSION THREAD

**Discussion by sam-wheat** [AvaloniaUI/Avalonia#8359] [Q&A] [answered]
Desired:  
	:is(BaseCell) styling will apply to all classes derived from BaseCell
	Background for TextBlock for MeasureCell will be Blue.

Actual:
	Background for TextBlock for MeasureCell is Red.

If there is a simpler/better way please point me in the right direction.  In WPF I used BasedOn in the styles which seems easier.

Classes:

	public class TotalCell : MeasureCell
	{
	}
	public class MeasureCell : BaseCell
	{
	}
	public class BaseCell : TemplatedControl
	{
	}
	
Styles:

	<Style Selector=":is(controls|BaseCell)">
		<Setter Property="Template">
			<ControlTemplate>
				<TextBlock Text="Base Cell" Background="Red" />
			</ControlTemplate>
		</Setter>
	</Style>

	<Style Selector="controls|MeasureCell /template/ TextBlock">
		<Setter Property="Background" Value="Blue" />
	</Style>

	<Style Selector="controls|MeasureCell TextBlock">
		<Setter Property="Background" Value="Blue" />
	</Style>

	<Style Selector="controls|MeasureCell > TextBlock">
		<Setter Property="Background" Value="Blue" />
	</Style>
	

[AvaloniaApp2.zip](https://github.com/AvaloniaUI/Avalonia/files/8934292/AvaloniaApp2.zip)

**Comment by jp2masa** [comment, marked as answer]
The problem here is that `Background="Red"` is being set with local value priority (which is the max priority I think), so the `Setter`s have no effect. You can solve it like this for example:

```xaml
<Style Selector=":is(controls|BaseCell)">
	<Setter Property="Template">
		<ControlTemplate>
			<TextBlock Text="Base Cell" />
		</ControlTemplate>
	</Setter>
</Style>

<Style Selector=":is(controls|BaseCell) /template/ TextBlock">
	<Setter Property="Background" Value="Red" />
</Style>
```

**Reply by sam-wheat** [reply]
Thanks @jp2masa.

**Reply by sam-wheat** [reply]
@jp2masa Can you provide a bit of insight as to why your answer works?

**Reply by jp2masa** [reply]
Maybe the docs can help: https://docs.avaloniaui.net/docs/styling/troubleshooting#locally-set-properties-override-styles.

The idea is that with my code the `Background` property will be set in a `Setter`, i.e. with the same priority as the other `Setter`s you already have. It's not set in the `TextBlock` directly, which would have higher priority and override all `Setter` values.

**Reply by sam-wheat** [reply]
> which would have higher priority and override all Setter values.

OK I see the problem now but I don't see how the styles are applied consistently.  I modified the XAML a bit:

    <Style Selector=":is(controls|BaseCell)">
		<Setter Property="Template">
			<ControlTemplate>
				<TextBlock Text="Default value" Background="Green"  />
			</ControlTemplate>
		</Setter>
	</Style>

	<Style Selector="controls|BaseCell TextBlock">
		<Setter Property="Background" Value="Red" />
		<Setter Property="Text" Value="Base Cell" />
	</Style>

	<Style Selector="controls|MeasureCell TextBlock">
		<Setter Property="Background" Value="Blue" />
		<Setter Property="Text" Value="Measure Cell" />
	</Style>

	<Style Selector="controls|TotalCell TextBlock">
		<Setter Property="Background" Value="Orange" />
		
	</Style>


In this case the `Background` and `Text` properties are set in the Textbox directly.  At runtime the `Text` property is overridden by the style more closely matching the type, whereas the `Background` property is not affected. 



![Capture](https://user-images.githubusercontent.com/16885210/174461107-6bad44a1-dbe7-496e-b35a-5ce159d6b41b.PNG)

**Reply by jp2masa** [reply]
That's a different question. Actually, you shouldn't use `Setter`s for `TextBlock.Text`, as it's a direct property, and you should only use `Setter`s for styled properties (see #7982 for more information).

**Comment by sam-wheat** [comment]
Yes, different question – sorry.  I am just getting started and going all over the place.  What a minefield.
